### ISSUE TITLE
typechecking does not terminate on contrived recursive type

### ISSUE THREAD

**Issue by fictitious** [microsoft/TypeScript#27073]
**TypeScript Version:** 3.1.0-dev.20180913

**Search Terms:**

typechecking infinite recursion

**Code**

```ts


export type DirectoryBase = {name: string}

export type DirectoryEntries = {readonly [n in string]: Directory};

export type Directory = DirectoryBase & {entries: DirectoryEntries};

export type ExactDirectory<E extends DirectoryEntries> = DirectoryBase & {entries: E};


export type Accessor<D extends Directory> = 
    D extends Directory ? {[K in keyof D['entries']]: Accessor<D['entries'][K]>} : {};


// intersection and conditional type are necessary to exhibit the problem
export type BadAccessor<D extends Directory> = {

} & (D extends Directory ? {[K in keyof D['entries']]: Accessor<D['entries'][K]>} : {});



function f<E extends DirectoryEntries, C extends Directory>(accessor: Accessor<C>): void {

    const a1 = accessor as Accessor<ExactDirectory<E>>; // ok, error as expected

    const a2 = accessor as BadAccessor<ExactDirectory<E>>; // boom
}

```

**Expected behavior:**

compilation error is reported

**Actual behavior:**

apparently, compilation does not terminate

**Playground Link:**  nope  **WARNING** pasting the above code to the playground freezes the playground UI, and because it remembers the code you are stuck forever. Browsers start warning about unresponsive page after a while.

**Related Issues:** 
might be the same as https://github.com/Microsoft/TypeScript/issues/27066 but observed behavior is different

**Comment by RyanCavanaugh** [comment]
I'm assuming "contrived" here means this what at least somewhat constructed instead of being "found" in real code - let me know if that's not the case. With Turing completeness sometimes comes nontermination 🤷‍♂️

**Comment by jakebailey** [comment]
I've been poking around old ancient issues; this looks to have been fixed in TS 3.3 (with an exact bisect to #29338).
